x-pack/libbeat/common/aws/credentials.go

```diff
@@ -44,134 +42,172 @@ type ConfigAWS struct {
 // If access keys given, use them as credentials.
 // If access keys are not given, then load from AWS config file. If credential_profile_name is not
 // given, default profile will be used.
 // If role_arn is given, assume the IAM role either with access keys or default profile.
-func GetAWSCredentials(config ConfigAWS) (awssdk.Config, error) {
+func GetAWSCredentials(beatsConfig ConfigAWS) (awssdk.Config, error) {
 	// Check if accessKeyID or secretAccessKey or sessionToken is given from configuration
-	if config.AccessKeyID != "" || config.SecretAccessKey != "" || config.SessionToken != "" {
-		return getAccessKeys(config), nil
+	if beatsConfig.AccessKeyID != "" || beatsConfig.SecretAccessKey != "" || beatsConfig.SessionToken != "" {
+		return getConfigForKeys(beatsConfig), nil
 	}
 
-	return getSharedCredentialProfile(config)
+	return getConfigSharedCredentialProfile(beatsConfig)
 }
 
-func getAccessKeys(config ConfigAWS) awssdk.Config {
-	awsConfig := defaults.Config()
+// getConfigForKeys creates a default AWS config and adds a CredentialsProvider using the provided Beats config.
+// Provided config must contain an accessKeyID, secretAccessKey and sessionToken to generate a valid CredentialsProfile
+func getConfigForKeys(beatsConfig ConfigAWS) awssdk.Config {
+	logger := logp.NewLogger("getConfigForKeys")
+
+	config := awssdk.NewConfig()
 	awsCredentials := awssdk.Credentials{
-		AccessKeyID:     config.AccessKeyID,
-		SecretAccessKey: config.SecretAccessKey,
+		AccessKeyID:     beatsConfig.AccessKeyID,
+		SecretAccessKey: beatsConfig.SecretAccessKey,
 	}
 
-	if config.SessionToken != "" {
-		awsCredentials.SessionToken = config.SessionToken
+	if beatsConfig.SessionToken != "" {
+		awsCredentials.SessionToken = beatsConfig.SessionToken
 	}
 
-	awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+	config.Credentials = credentials.StaticCredentialsProvider{
 		Value: awsCredentials,
 	}
 
-	return awsConfig
+	// Assume IAM role if iam_role config parameter is given
+	if beatsConfig.RoleArn != "" {
+		logger.Debug("Using role arn and access keys for AWS credential")
+		addStaticCredentialsProviderToAwsConfig(beatsConfig, config)
+		return *config
+	}
+
+	return *config
 }
 
-func getSharedCredentialProfile(config ConfigAWS) (awssdk.Config, error) {
-	// If accessKeyID, secretAccessKey or sessionToken is not given, iam_role is not given, then load from default config
-	// Please see https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html
-	// with more details.
-	// If credential_profile_name is empty, then default profile is used.
-	logger := logp.NewLogger("getSharedCredentialProfile")
-	var options []external.Config
-	if config.ProfileName != "" {
-		options = append(options, external.WithSharedConfigProfile(config.ProfileName))
+// getConfigSharedCredentialProfile If accessKeyID, secretAccessKey or sessionToken is not given, iam_role is not given,
```

Comment:
the comment is misleading: we still jump to this function even if `iam_role` is given

Reply:
Yep I just removed the `iam_role` part from the comment. With given `iam_role`, we still need to have either the credential profile or access keys first.